x-pack/plugins/endpoint/server/routes/resolver/utils/tree.ts

```diff
@@ -0,0 +1,211 @@
+//
+// Following this scheme, we use exactly 2 queries per level of children that we return--one for the pagination
+// and one for the lifecycle events of the processes. The downside to this is that we need to dynamically expand
+// the number of documents we can retrieve per level due to the exponential fanout of child processes,
+// what this means is that noisy neighbors for a given level may hide other child process events that occur later
+// temporally in the same level--so, while a heavily forking process might get shown, maybe the actually malicious
+// event doesn't show up in the tree at the beginning.
+export class Tree {
+  protected cache: Record<string, Node>;
+  protected root: Node;
+  protected id: string;
+
+  constructor(id: string) {
+    const root = createNode(id);
+    this.id = id;
+    this.cache = { [id]: root };
+    this.root = root;
+  }
+
+  public render(): Node {
+    return this.root;
+  }
+
+  public ids(): string[] {
+    return Object.keys(this.cache);
+  }
+
+  public static async merge(
+    childrenPromise: Promise<Tree>,
+    ancestorsPromise: Promise<Tree>,
+    eventsPromise: Promise<Tree>
+  ): Promise<Tree> {
+    const [children, ancestors, events] = await Promise.all([
+      childrenPromise,
+      ancestorsPromise,
+      eventsPromise,
+    ]);
+
+    // we only allow for merging when we have partial trees that
+    // represent the same root node
+    const rootID = children.id;
+    if (rootID !== ancestors.id || rootID !== events.id) {
+      throw new Error('cannot merge trees with different roots');
+    }
+
+    // our caches should be exclusive for everything but the root node so we can
+    // just merge them
+    Object.entries(ancestors.cache).forEach(([id, node]) => {
+      if (rootID !== id) children.cache[id] = node;
+    });
+
+    // fix up the references
+    children.root.lifecycle = ancestors.root.lifecycle; // lifecycle is bound to the ancestors query
+    children.root.parent = ancestors.root.parent;
+    children.root.ancestors = ancestors.root.ancestors;
+    children.root.events = events.root.events;
+
+    // merge the pagination
+    Object.assign(children.root.pagination, ancestors.root.pagination, events.root.pagination);
+
+    return children;
+  }
+
+  public addEvent(...events: ResolverEvent[]): void {
+    events.forEach(event => {
+      const id = entityId(event);
+
+      this.ensureCache(id);
+      this.cache[id].events.push(event);
+    });
+  }
+
+  public addAncestor(id: string, ...events: ResolverEvent[]): void {
+    events.forEach(event => {
+      const ancestorID = entityId(event);
+      if (!this.cache[ancestorID]) {
+        const newParent = createNode(ancestorID);
+        this.cache[ancestorID] = newParent;
+        // Uncomment this line to populate the node's parent field, it will build a tree like
+        // node.parent.parent.parent going up the ancestry chain
+        // this.cache[id].parent = newParent;
```

Comment:
@kqualters-elastic should we remove the `parent` field from `Node` and get rid of it here? The UI is only using `ancestors` right?